session/cache/session.go

```diff
@@ -0,0 +1,353 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/user"
+	"path/filepath"
+
+	"github.com/vmware/govmomi/session"
+	"github.com/vmware/govmomi/vapi/rest"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// Client interface to support client session caching
+type Client interface {
+	json.Marshaler
+	json.Unmarshaler
+
+	Valid() bool
+	Path() string
+}
+
+// Session provides methods to cache authenticated vim25.Client and rest.Client sessions.
+// Use of session cache avoids the expense of creating and deleting vSphere sessions.
+// It also helps avoid the problem of "leaking sessions", as Session.Login will only
+// create a new authenticated session if the cached session does not exist or is invalid.
+// By default, username/password authentication is used to create new sessions.
+// The Session.Login{SOAP,REST} fields can be set to use other methods,
+// such as SAML token authentication (see govc session.login for example).
+type Session struct {
+	URL         *url.URL // URL of a vCenter or ESXi instance
+	Dir         string   // Dir defaults to "$HOME/.govmomi/sessions"
+	Insecure    bool     // Insecure param for soap.NewClient (tls.Config.InsecureSkipVerify)
+	Passthrough bool     // Passthrough disables caching when set to true
+
+	LoginSOAP func(context.Context, *vim25.Client) error // LoginSOAP defaults to session.Manager.Login()
+	LoginREST func(context.Context, *rest.Client) error  // LoginREST defaults to rest.Client.Login()
+}
+
+var (
+	home = os.Getenv("GOVMOMI_HOME")
+)
+
+func init() {
+	if home == "" {
+		dir, err := os.UserHomeDir()
+		if err != nil {
+			dir = os.Getenv("HOME")
+		}
+		home = filepath.Join(dir, ".govmomi")
+	}
+}
+
+// Endpoint returns a copy of the Session.URL with Password, Query and Fragment removed.
+func (s *Session) Endpoint() *url.URL {
+	if s.URL == nil {
+		return nil
+	}
+	p := &url.URL{
+		Scheme: s.URL.Scheme,
+		Host:   s.URL.Host,
+		Path:   s.URL.Path,
+	}
+	if u := s.URL.User; u != nil {
+		p.User = url.User(u.Username()) // Remove password
+	}
+	return p
+}
+
+// key is a digest of the URL scheme + username + host + Client.Path()
+func (s *Session) key(path string) string {
+	p := s.Endpoint()
+	p.Path = path
+
+	// Key session file off of full URI and insecure setting.
+	// Hash key to get a predictable, canonical format.
+	key := fmt.Sprintf("%s#insecure=%t", p.String(), s.Insecure)
+	return fmt.Sprintf("%064x", sha256.Sum256([]byte(key)))
+}
+
+func (s *Session) dir() string {
```

Comment:
Considering [we are trying to match Terraform's persisted REST session location](https://github.com/vmware/govmomi/issues/1810#issuecomment-578351354). Should the persisted REST session not be stored in:

> ${HOME}/.govmomi/rest_sessions

As apposed to where the SOAP sessions are stored:

> ${HOME}/.govmomi/sessions

The [Terraform vsphere_provider seems to store them as such](https://www.terraform.io/docs/providers/vsphere/index.html#session-persistence-options). @dougm?

Reply:
Thanks for reviewing @matthewmfrost .  We could use separate directories, but is it useful to have them separate?  Note that the path (`/soap` or `/rest`) is included in the hashed file name, so the file names will be unique within the same directory.

Reply:
I completely agree that is doesn't gain that much having them separate directories. I am not sure why the [terraform-provider-vsphere](https://github.com/terraform-providers/terraform-provider-vsphere) chose to do it that way, other than a bit more clarity for session management debugging. 

But it would be **nice** to be consistent given that the two (govc and terraform) are often used together. The GOVC/Terraform session interoperability is really neat on the SOAP side!

Reply:
Agreed, having them consistent is a must for sharing the session cache.
It's probably less trouble to change the govmomi side since this is a new package.  I'll go ahead and do that later today.

Reply:
Great, looks good to me :)